### PR TITLE
gitserver: grpc: create FirstEverCommit RPC method and server implementation

### PR DESCRIPTION
**Pull request by ggilmore**
Part of #61689 

This PR builds on https://github.com/sourcegraph/sourcegraph/pull/62165, and adds the FirstEverCommit new gRPC method along with a corresponding server implementation. The client implementation will be in a future PR.

## Test plan

- Unit tests

### CODE REVIEW

**Comment by ggilmore** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62169?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#62169** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62169?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#62165** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#62158** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62158?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ggilmore and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: cmd/gitserver/internal/server_grpc.go [hunk 1]
```diff
@@ -1408,6 +1408,60 @@ func (gs *grpcServer) ContributorCounts(ctx context.Context, req *proto.Contribu
...
+		if errors.As(err, &revisionErr) {
+			s, err := status.New(codes.NotFound, "revision not found").WithDetails(&proto.RevisionNotFoundPayload{
+				Repo: request.GetRepoName(),
+				Spec: revisionErr.Spec,
+			})
+			if err != nil {
+				return nil, err
+			}
+			return nil, s.Err()
+		}
```
Comment (eseliger) [inline cmd/gitserver/internal/server_grpc.go:1454]: do we currently return a revision not found error? (see other PR, I wonder if we should)

File: cmd/gitserver/internal/server_grpc_test.go [hunk 2]
```diff
@@ -1064,6 +1064,102 @@ func TestGRPCServer_ContributorCounts(t *testing.T) {
 	})
 }
 
+func TestGRPCServer_FirstCommitEver(t *testing.T) {
+	// Add an actor to the context.
+	a := actor.FromUser(1)
+	ctx := actor.WithActor(context.Background(), a)
```
Comment (eseliger) [inline cmd/gitserver/internal/server_grpc_test.go:1070]: Do we need an actor? FYI, recently removed subrepo perms from the server-side again

File: internal/gitserver/v1/gitserver.proto [hunk 3]
```diff
@@ -228,6 +228,14 @@ service GitserverService {
...
     option idempotency_level = NO_SIDE_EFFECTS;
   }
+
+  // FirstEverCommit returns the first commit ever made to the repository.
+  //
+  // If the given repository is empty, an error with a RepoNotFoundPayload is
+  // returned.
+  rpc FirstEverCommit(FirstEverCommitRequest) returns (FirstEverCommitResponse) {
+    option idempotency_level = NO_SIDE_EFFECTS;
+  }
```
Comment (eseliger) [inline internal/gitserver/v1/gitserver.proto:238]: we should document the `RepositoryEmptyPayload` case here

**Review by eseliger — APPROVED** [review]

**Comment by ggilmore** [comment]
### Merge activity

* **May 1, 12:16 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/62169) rebased this pull request as part of a merge.

**Pull request merged**